File: mobile/lib/services/financial_engine.dart
import 'dart:math';
import 'package:intl/intl.dart';
import '../models/transaction_item.dart';
import '../models/cash_flow_summary.dart';
import '../models/insight_item.dart';

enum DemoScenario {
  personalNormal,
  personalHighSpending,
  personalLowBalance,
  smeGrowingRevenue,
  smeRisingExpenses,
}

class FinancialEngine {
  static final NumberFormat currencyFormatter = NumberFormat.currency(
    symbol: '₦',
    decimalDigits: 0,
  );

  static String formatCurrency(double amount) {
    return currencyFormatter.format(amount);
  }

  static CashFlowSummary calculateSummary({
    required List<TransactionItem> transactions,
    required bool isBusinessMode,
    DemoScenario scenario = DemoScenario.personalNormal,
  }) {
    if (transactions.isEmpty) {
      return CashFlowSummary(
        currentBalance: 0.0,
        moneyIn: 0.0,
        moneyOut: 0.0,
        netCashFlow: 0.0,
        moneyInTrendPercent: 0.0,
        moneyOutTrendPercent: 0.0,
        projectedBalance30D: 0.0,
        runwayDays: 0,
        chartPoints: [],
        categoryBreakdowns: [],
      );
    }

    // Sort ascending by date
    final sorted = List<TransactionItem>.from(transactions)
      ..sort((a, b) => a.date.compareTo(b.date));

    double moneyIn = 0.0;
    double moneyOut = 0.0;
    final Map<String, double> categoryAmounts = {};
    final Map<String, int> categoryCounts = {};

    for (final tx in sorted) {
      final amt = tx.amount.abs();
      if (tx.type == TransactionType.credit) {
        moneyIn += amt;
      } else {
        moneyOut += amt;
        final cat = tx.category;
        categoryAmounts[cat] = (categoryAmounts[cat] ?? 0.0) + amt;
        categoryCounts[cat] = (categoryCounts[cat] ?? 0) + 1;
      }
    }

    final double netCashFlow = moneyIn - moneyOut;

    // Use latest transaction balance if positive, otherwise netCashFlow
    final latestTx = sorted.last;
    final double currentBalance = (latestTx.balanceAfter > 0)
        ? latestTx.balanceAfter
        : max(0.0, netCashFlow);

    // Build category breakdowns
    final List<CategoryBreakdownItem> categoryBreakdowns = [];
    if (moneyOut > 0) {
      for (final entry in categoryAmounts.entries) {
        final pct = ((entry.value / moneyOut) * 1000).round() / 10.0;
        categoryBreakdowns.add(CategoryBreakdownItem(
          category: entry.key,
          amount: entry.value,
          percentage: pct,
          count: categoryCounts[entry.key] ?? 1,
        ));
      }
      categoryBreakdowns.sort((a, b) => b.amount.compareTo(a.amount));
    }

    // Group chart points by month
    final Map<String, Map<String, double>> monthMap = {};
    final DateFormat monthFormat = DateFormat('MMM');

    for (final tx in sorted) {
      final key = monthFormat.format(tx.date);
      final current = monthMap.putIfAbsent(key, () => {'income': 0.0, 'expense': 0.0});
      if (tx.type == TransactionType.credit) {
        current['income'] = (current['income'] ?? 0.0) + tx.amount.abs();
      } else {
        current['expense'] = (current['expense'] ?? 0.0) + tx.amount.abs();
      }
    }

    final List<CashFlowPoint> chartPoints = [];
    double runningBalance = 0.0;

    for (final entry in monthMap.entries) {
      final inc = entry.value['income'] ?? 0.0;
      final exp = entry.value['expense'] ?? 0.0;
      final net = inc - exp;
      runningBalance += net;

      chartPoints.add(CashFlowPoint(
        dateLabel: entry.key,
        income: inc,
        expense: exp,
        net: net,
        balance: runningBalance,
        isForecast: false,
      ));
    }

    // Dynamic runway & 30-day forecast
    int runwayDays = 0;
    double projectedBalance30D = currentBalance;

    if (sorted.length > 1) {
      final int totalDays = max(1, sorted.last.date.difference(sorted.first.date).inDays);
      final double dailyBurn = moneyOut / totalDays;
      final double dailyIncome = moneyIn / totalDays;

      if (dailyBurn > 0) {
        runwayDays = max(1, (currentBalance / dailyBurn).round());
      } else {
        runwayDays = 90;
      }

      final projectedOut = dailyBurn * 30;
      final projectedIn = dailyIncome * 30;
      projectedBalance30D = max(0.0, currentBalance + projectedIn - projectedOut);

      if (chartPoints.isNotEmpty) {
        chartPoints.add(CashFlowPoint(
          dateLabel: 'Next 30D (Proj)',
          income: projectedIn,
          expense: projectedOut,
          net: projectedIn - projectedOut,
          balance: projectedBalance30D,
          isForecast: true,
        ));
      }
    }

    return CashFlowSummary(
      currentBalance: currentBalance,
      moneyIn: moneyIn,
      moneyOut: moneyOut,
      netCashFlow: netCashFlow,
      moneyInTrendPercent: 0.0,
      moneyOutTrendPercent: 0.0,
      projectedBalance30D: projectedBalance30D,
      runwayDays: runwayDays,
      chartPoints: chartPoints,
      categoryBreakdowns: categoryBreakdowns,
    );
  }

  static List<InsightItem> generateInsights({
    required List<TransactionItem> transactions,
    required CashFlowSummary summary,
    required bool isBusinessMode,
  }) {
    if (transactions.isEmpty) {
      return [];
    }

    final List<InsightItem> insights = [];

    // 1. Cash flow direction
    if (summary.netCashFlow >= 0) {
      insights.add(InsightItem(
        id: 'ins-pos-flow',
        type: InsightType.positiveCashFlow,
        title: 'Positive Cash Flow',
        description:
            'Total inflows (${formatCurrency(summary.moneyIn)}) outpaced outflows (${formatCurrency(summary.moneyOut)}) by ${formatCurrency(summary.netCashFlow)}.',
        severity: InsightSeverity.positive,
        metric: '+${formatCurrency(summary.netCashFlow)}',
        comparison: 'Net surplus',
        date: DateTime.now(),
      ));
    } else {
      insights.add(InsightItem(
        id: 'ins-neg-flow',
        type: InsightType.spendingIncrease,
        title: 'Outflows Exceeded Inflows',
        description:
            'Total spending (${formatCurrency(summary.moneyOut)}) was greater than income (${formatCurrency(summary.moneyIn)}) by ${formatCurrency(summary.netCashFlow.abs())}.',
        severity: InsightSeverity.warning,
        metric: '-${formatCurrency(summary.netCashFlow.abs())}',
        comparison: 'Deficit',
        date: DateTime.now(),
        ctaText: 'Review expenses',
        ctaRoute: '/transactions',
      ));
    }

    // 2. Top spending category
    if (summary.categoryBreakdowns.isNotEmpty) {
      final topCat = summary.categoryBreakdowns.first;
      insights.add(InsightItem(
        id: 'ins-top-cat',
        type: InsightType.highExpenseCategory,
        title: 'Top Spending: ${topCat.category}',
        description:
            '${topCat.category} represents ${topCat.percentage}% of your total expenses (${formatCurrency(topCat.amount)} across ${topCat.count} transactions).',
        severity: topCat.percentage > 35 ? InsightSeverity.warning : InsightSeverity.info,
        metric: '${topCat.percentage}%',
        comparison: formatCurrency(topCat.amount),
        date: DateTime.now(),
        ctaText: 'View transactions',
        ctaRoute: '/transactions',
      ));
    }

    // 3. Runway Status
    if (summary.runwayDays > 0) {
      final isLow = summary.runwayDays < 15;
      insights.add(InsightItem(
        id: 'ins-runway',
        type: InsightType.lowBalanceForecast,
        title: isBusinessMode ? 'Operating Runway Status' : 'Cash Runway Estimate',
        description:
            'Based on recent daily outflow rate, your liquid balance provides approximately ${summary.runwayDays} days of coverage.',
        severity: isLow ? InsightSeverity.critical : InsightSeverity.info,
        metric: '${summary.runwayDays} Days',
        comparison: 'Estimated runway',
        date: DateTime.now(),
        ctaText: 'View cash-flow forecast',
        ctaRoute: '/cash-flow',
      ));
    }

    return insights;
  }
}

File: web/src/services/FinancialEngine.ts
import { CashFlowSummary, CashFlowPoint, CategoryBreakdown, Insight, Transaction } from '../types/financial';

export class FinancialEngine {
  static formatNaira(amount: number): string {
    const isNegative = amount < 0;
    const abs = Math.abs(Math.round(amount));
    const formatted = abs.toLocaleString('en-NG');
    return `${isNegative ? '-' : ''}₦${formatted}`;
  }

  static calculateSummary(transactions: Transaction[], isBusiness: boolean = false): CashFlowSummary {
    if (!transactions || transactions.length === 0) {
      return {
        currentBalance: 0,
        moneyIn: 0,
        moneyOut: 0,
        netCashFlow: 0,
        moneyInTrendPercent: 0,
        moneyOutTrendPercent: 0,
        projectedBalance30D: 0,
        runwayDays: 0,
        outstandingCommitments: 0,
        chartPoints: [],
        categoryBreakdowns: [],
      };
    }

    // Sort transactions by date ascending
    const sorted = [...transactions].sort(
      (a, b) => new Date(a.date).getTime() - new Date(b.date).getTime()
    );

    let moneyIn = 0;
    let moneyOut = 0;
    const categoryMap = new Map<string, { amount: number; count: number }>();

    for (const tx of sorted) {
      const amt = Math.abs(tx.amount);
      if (tx.type === 'credit') {
        moneyIn += amt;
      } else {
        moneyOut += amt;
        const cat = tx.category || 'Other';
        const existing = categoryMap.get(cat) || { amount: 0, count: 0 };
        existing.amount += amt;
        existing.count += 1;
        categoryMap.set(cat, existing);
      }
    }

    const netCashFlow = moneyIn - moneyOut;

    // Current balance: use the latest transaction's balanceAfter if valid, otherwise net cash flow
    const latestTx = sorted[sorted.length - 1];
    const currentBalance =
      latestTx.balanceAfter !== undefined && latestTx.balanceAfter !== null
        ? latestTx.balanceAfter
        : Math.max(0, netCashFlow);

    // Category breakdowns
    const categoryBreakdowns: CategoryBreakdown[] = [];
    if (moneyOut > 0) {
      for (const [category, data] of categoryMap.entries()) {
        const percentage = Math.round((data.amount / moneyOut) * 1000) / 10;
        categoryBreakdowns.push({
          category,
          amount: data.amount,
          percentage,
          count: data.count,
        });
      }
      categoryBreakdowns.sort((a, b) => b.amount - a.amount);
    }

    // Chart points: Group by month
    const monthMap = new Map<string, { income: number; expense: number }>();
    for (const tx of sorted) {
      const d = new Date(tx.date);
      const monthKey = !isNaN(d.getTime())
        ? `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}`
        : '2026-01';
      const current = monthMap.get(monthKey) || { income: 0, expense: 0 };
      if (tx.type === 'credit') {
        current.income += Math.abs(tx.amount);
      } else {
        current.expense += Math.abs(tx.amount);
      }
      monthMap.set(monthKey, current);
    }

    const monthNames = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
    const chartPoints: CashFlowPoint[] = [];
    let runningBalance = 0;

    const sortedMonthKeys = Array.from(monthMap.keys()).sort();
    for (const key of sortedMonthKeys) {
      const data = monthMap.get(key)!;
      const [, m] = key.split('-');
      const monthIndex = parseInt(m, 10) - 1;
      const dateLabel = monthNames[monthIndex] || key;
      const net = data.income - data.expense;
      runningBalance += net;

      chartPoints.push({
        dateLabel,
        income: data.income,
        expense: data.expense,
        net,
        balance: runningBalance,
        isForecast: false,
      });
    }

    // Dynamic runway & 30-day forecast
    let runwayDays = 0;
    let projectedBalance30D = currentBalance;

    if (sorted.length > 1) {
      const firstDate = new Date(sorted[0].date).getTime();
      const lastDate = new Date(sorted[sorted.length - 1].date).getTime();
      const totalDays = Math.max(1, Math.round((lastDate - firstDate) / (1000 * 60 * 60 * 24)));
      const dailyBurn = moneyOut / totalDays;
      const dailyIncome = moneyIn / totalDays;

      if (dailyBurn > 0) {
        runwayDays = Math.max(1, Math.round(currentBalance / dailyBurn));
      } else {
        runwayDays = 90;
      }

      const projectedOut = dailyBurn * 30;
      const projectedIn = dailyIncome * 30;
      projectedBalance30D = Math.max(0, Math.round(currentBalance + projectedIn - projectedOut));

      // Add projected point to chart if we have at least 1 month
      if (chartPoints.length > 0) {
        chartPoints.push({
          dateLabel: 'Next 30D (Proj)',
          income: Math.round(projectedIn),
          expense: Math.round(projectedOut),
          net: Math.round(projectedIn - projectedOut),
          balance: projectedBalance30D,
          isForecast: true,
        });
      }
    }

    return {
      currentBalance,
      moneyIn,
      moneyOut,
      netCashFlow,
      moneyInTrendPercent: 0,
      moneyOutTrendPercent: 0,
      projectedBalance30D,
      runwayDays,
      outstandingCommitments: 0,
      chartPoints,
      categoryBreakdowns,
    };
  }

  static generateInsights(transactions: Transaction[], summary: CashFlowSummary, isBusiness: boolean = false): Insight[] {
    if (!transactions || transactions.length === 0) {
      return [];
    }

    const insights: Insight[] = [];

    // 1. Cash flow direction
    if (summary.netCashFlow >= 0) {
      insights.push({
        id: 'ins-pos-flow',
        type: 'positive_cash_flow',
        title: 'Positive Cash Flow',
        description: `Your total inflows (${this.formatNaira(summary.moneyIn)}) exceeded your outflows (${this.formatNaira(summary.moneyOut)}) by ${this.formatNaira(summary.netCashFlow)}.`,
        severity: 'positive',
        metric: `+${this.formatNaira(summary.netCashFlow)}`,
        comparison: 'Net surplus',
        date: new Date().toISOString().split('T')[0],
      });
    } else {
      insights.push({
        id: 'ins-neg-flow',
        type: 'spending_increase',
        title: 'Outflows Exceeded Inflows',
        description: `Your total spending (${this.formatNaira(summary.moneyOut)}) was greater than income (${this.formatNaira(summary.moneyIn)}) by ${this.formatNaira(Math.abs(summary.netCashFlow))}.`,
        severity: 'warning',
        metric: `-${this.formatNaira(Math.abs(summary.netCashFlow))}`,
        comparison: 'Deficit',
        date: new Date().toISOString().split('T')[0],
        ctaText: 'Review expenses',
        ctaRoute: '/transactions',
      });
    }

    // 2. Top spending category
    if (summary.categoryBreakdowns.length > 0) {
      const topCat = summary.categoryBreakdowns[0];
      insights.push({
        id: 'ins-top-cat',
        type: 'high_expense_category',
        title: `Top Spending: ${topCat.category}`,
        description: `${topCat.category} represents ${topCat.percentage}% of your total expenses (${this.formatNaira(topCat.amount)} across ${topCat.count} transactions).`,
        severity: topCat.percentage > 35 ? 'warning' : 'info',
        metric: `${topCat.percentage}%`,
        comparison: this.formatNaira(topCat.amount),
        date: new Date().toISOString().split('T')[0],
        ctaText: 'View transactions',
        ctaRoute: '/transactions',
      });
    }

    // 3. Runway Status
    if (summary.runwayDays > 0) {
      const isLow = summary.runwayDays < 15;
      insights.push({
        id: 'ins-runway',
        type: 'low_balance_forecast',
        title: isBusiness ? 'Operating Runway Status' : 'Cash Runway Estimate',
        description: `Based on your recent average daily burn rate, your current balance provides approximately ${summary.runwayDays} days of coverage.`,
        severity: isLow ? 'critical' : 'info',
        metric: `${summary.runwayDays} Days`,
        comparison: 'Estimated runway',
        date: new Date().toISOString().split('T')[0],
        ctaText: 'View cash-flow forecast',
        ctaRoute: '/cash-flow',
      });
    }

    return insights;
  }
}
